From 883860d245aa089364e9832da08f1a76987b17eb Mon Sep 17 00:00:00 2001
From: Ori Livneh <ori@wikimedia.org>
Date: Sat, 30 Aug 2014 17:49:33 -0700
Subject: [PATCH] Make RUSAGE_THREAD available to getrusage PHP function

The signature of the PHP function is getrusage( [ int $who = 0 ] ).
In both HHVM and PHP5, $who can be 1 to request RUSAGE_CHILDREN data. This
change allows $who to be 2 to request RUSAGE_THREAD data. RUSAGE_THREAD
measures resource usage for just the calling thread.

This allows HHVM PHP code to measure resource usage for the current request.
(In PHP5 this can be done with RUSAGE_SELF, by dint of the threading model.)
---
 hphp/runtime/ext/std/ext_std_options.cpp | 18 +++++++++++++++++-
 1 file changed, 17 insertions(+), 1 deletion(-)

diff --git a/hphp/runtime/ext/std/ext_std_options.cpp b/hphp/runtime/ext/std/ext_std_options.cpp
index 4b9a84c..369b45f 100644
--- a/hphp/runtime/ext/std/ext_std_options.cpp
+++ b/hphp/runtime/ext/std/ext_std_options.cpp
@@ -709,8 +709,24 @@ const StaticString
 static Array HHVM_FUNCTION(getrusage, int64_t who /* = 0 */) {
   struct rusage usg;
   memset(&usg, 0, sizeof(struct rusage));
+  int actual_who;
+  switch (who) {
+  case 1:
+    actual_who = RUSAGE_CHILDREN;
+    break;
+  case 2:
+#ifdef RUSAGE_THREAD
+    actual_who = RUSAGE_THREAD;
+#else
+    throw_not_supported(__func__, "RUSAGE_THREAD is not defined on this sytem");
+#endif
+    break;
+  default:
+    actual_who = RUSAGE_SELF;
+    break;
+  }
 
-  if (getrusage(who == 1 ? RUSAGE_CHILDREN : RUSAGE_SELF, &usg) == -1) {
+  if (getrusage(actual_who, &usg) == -1) {
     raise_error("getrusage returned %d: %s", errno,
       folly::errnoStr(errno).c_str());
   }
